src/v/kafka/server/handlers/alter_partition_reassignments.cc

```diff
@@ -0,0 +1,151 @@
+        return error_code::invalid_topic_exception;
+    default:
+        return error_code::unknown_server_error;
+    }
+}
+
+template<>
+ss::future<response_ptr> alter_partition_reassignments_handler::handle(
+  request_context ctx, [[maybe_unused]] ss::smp_service_group ssg) {
+    alter_partition_reassignments_request request;
+    request.decode(ctx.reader(), ctx.header().version);
+    log_request(ctx.header(), request);
+    alter_partition_reassignments_response resp;
+
+    if (request.data.topics.empty()) {
+        co_return co_await ctx.respond(std::move(resp));
+    }
+
+    resp.data.responses.reserve(request.data.topics.size());
+    auto timeout_ms = request.data.timeout_ms;
+    kafka::error_code top_level_err{kafka::error_code::none};
+
+    for (const reassignable_topic& topic : request.data.topics) {
+        reassignable_topic_response topic_response{.name = topic.name};
+
+        for (const reassignable_partition& partition : topic.partitions) {
+            model::ntp ntp{
+              model::ns{"kafka"}, topic.name, partition.partition_index};
+            if (partition.replicas.has_value()) {
+                // Convert each node id to a broker shard because
+                // topics_frontend.move_partition_replicas expects
+                // vector<broker_shard>
+                std::vector<model::broker_shard> replicas;
+                std::for_each(
+                  partition.replicas->cbegin(),
+                  partition.replicas->cend(),
+                  [&replicas](const model::node_id& node_id) {
+                      replicas.push_back(
+                        model::broker_shard{.node_id = node_id, .shard = 0});
+                  });
+
+                vlog(
+                  klog.debug,
+                  "Request to reassign partitions: ntp {}, replicas {}",
+                  ntp,
+                  replicas);
+
+                auto errc
+                  = co_await ctx.topics_frontend().move_partition_replicas(
+                    ntp, replicas, model::timeout_clock::now() + timeout_ms);
+
+                if (!errc) {
+                    topic_response.partitions.push_back(
+                      reassignable_partition_response{
+                        .partition_index = partition.partition_index});
+                } else {
+                    auto kerr = error_code_as_kafka_error(errc);
+                    vlog(
+                      klog.error,
+                      "Unknown problem moving partitions: ntp {}, ec {} - {}",
+                      ntp,
+                      kerr,
+                      kafka::error_code_to_str(kerr));
+                    topic_response.partitions.push_back(
+                      reassignable_partition_response{
+                        .partition_index = model::partition_id{-1},
+                        .error_code = kerr,
+                        .error_message = ss::sstring{
+                          kafka::error_code_to_str(kerr)}});
```

Comment:
There must be a cleaner way to capture errors and convert a `std::error_code` to a `kafka::error_code`